File: book/petrol-pump/chapter3/section5/NotifyPointOfSale.java
package chapter3.section5;

import pump.*;
import chapter3.section2.LifeCycle;
import chapter3.section2.LifeCycle.End;
import chapter3.section4.Fill;
import sodium.*;
import java.util.Optional;

public class NotifyPointOfSale {
    public final Event<Fuel> eStart;
    public final Behavior<Optional<Fuel>> fillActive;
    public final Behavior<Optional<Fuel>> fuelFlowing;
    public final Event<End> eEnd;
    public final Event<Unit> eBeep;
    public final Event<Sale> eSaleComplete;

    public NotifyPointOfSale(
             LifeCycle lc,
             Event<Unit> eClearSale,
             Fill fi) {
        Behavior<Boolean> locked =
                lc.eStart.map(u -> true).merge(
                eClearSale.map(u -> false)).hold(false);
        eStart = lc.eStart.gate(locked.map(l -> !l));
        eEnd    = lc.eEnd.gate(locked);
        fuelFlowing =
                eStart.map(f -> Optional.of(f)).merge(
                eEnd.map(f -> Optional.empty())).hold(Optional.empty());
        fillActive =
             eStart.map(f -> Optional.of(f)).merge(
             eClearSale.map(f -> Optional.empty())).hold(Optional.empty());
        eBeep = eClearSale;
        eSaleComplete = Event.filterOptional(eEnd.snapshot(
            Behavior.lift(
                (oFuel, price_, dollars, liters) ->
                    oFuel.isPresent() ? Optional.of(
                           new Sale(oFuel.get(), price_, dollars, liters))
                                      : Optional.empty(),
                fuelFlowing, fi.price, fi.dollarsDelivered,
                fi.litersDelivered)));
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter3/Section7/NotifyPointOfSale.cs
using System;

using PetrolPump.Chapter3.Section3;
using PetrolPump.Chapter3.Section6;
using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter3.Section7
{
  public class NotifyPointOfSale
  {
    public readonly Event<Fuel> EStart;
    public readonly Behavior<Optional<Fuel>> FillActive;
    public readonly Behavior<Optional<Fuel>> FuelFlowing;
    public readonly Event<LifeCycle.End> EEnd;
    public readonly Event<Unit> EBeep;
    public readonly Event<Sale> ESaleComplete;

    public NotifyPointOfSale(
      LifeCycle lc,
      Event<Unit> eClearSale,
      Fill fi)
    {
      Behavior<Boolean> locked = lc.EStart.Map(u => true).Merge(eClearSale.Map(u => false)).Hold(false);
      EStart = lc.EStart.Gate(locked.Map(l => !l));
      EEnd = lc.EEnd.Gate(locked);
      FuelFlowing = EStart.Map(f => Optional<Fuel>.Of(f))
        .Merge(EEnd.Map(f => Optional<Fuel>.Empty())).Hold(Optional<Fuel>.Empty());
      FillActive = EStart.Map(f => Optional<Fuel>.Of(f))
        .Merge(eClearSale.Map(f => Optional<Fuel>.Empty())).Hold(Optional<Fuel>.Empty());
      EBeep = eClearSale;
      ESaleComplete = Event<Sale>.FilterOptional(
        EEnd.Snapshot(
          Behavior<Sale>.Lift(
            (oFuel, price_, dollars, liters) => oFuel.IsPresent ? Optional<Sale>.Of(new Sale(oFuel.Get(), price_, dollars, liters)) : Optional<Sale>.Empty(),
            FuelFlowing,
            fi.Price,
            fi.DollarsDelivered,
            fi.LitersDelivered)));
    }
  }

}
